### PR TITLE
feat: UserDAO

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

We want to be able to persist entities locally, preferably without having to re-write the persistence logic for each platform.

### Solutions

- Integrate SQLDelight as a multi platform persistence solution (doesn't work for yet for web)
- Implement UserDAO for persisting users

### Testing

Wrote common tests against the UserDAO interface.

### Notes (Optional)

~iOS is currently broken since we don't use multithreaded coroutines.~

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by typfel** [comment]
> @typfel what happens on iOS? Because with Kotlin 1.6's new memory model + coroutines 1.6, we should be able to get some multi-threading on coroutines.
> 
> Other than that, seems pretty nice so far!

We currently we get a `kotlin.native.concurrent.InvalidMutabilityException` when using the flow extension. Discussed in this thread: https://github.com/cashapp/sqldelight/issues/1845

I did try to use the `native-mt` version of 1.6.0 but I ran to a lot of gradle errors when doing that.

**Comment by typfel** [comment]
> I did try to use the native-mt version of 1.6.0 but I ran into to a lot of gradle errors when doing that.

Today it worked and now the tests also pass on iOS/OSX

**Review by vitorhugods — APPROVED** [review]
🧑‍🍳🤌🏻

**Review by MohamadJaara — APPROVED** [review]
well done! 🚀

File: persistence/src/androidMain/kotlin/com/wire/kalium/persistence/db/Database.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.wire.kalium.persistence.db
+
+import android.content.Context
+import app.cash.sqldelight.driver.android.AndroidSqliteDriver
+import com.wire.kalium.persistence.dao.QualifiedIDAdapter
+import com.wire.kalium.persistence.dao.UserDAO
+import com.wire.kalium.persistence.dao.UserDAOImpl
+import net.sqlcipher.database.SupportFactory
+
+actual class Database(context: Context, name: String, passphrase: String) {
```
[the remark this answers is not in the document]
Comment (typfel) [reply persistence/src/androidMain/kotlin/com/wire/kalium/persistence/db/Database.kt:10]: Absolutely that makes a ton of sense. I didn't put much thought into this part of the API

[the remark this answers is not in the document]
Comment (vitorhugods) [reply persistence/src/androidMain/kotlin/com/wire/kalium/persistence/db/Database.kt:10]: With #188 we can generate and store the `passphrase` in a secure way without writing platform-specific code, right? I'm getting excited! Linking it all together!

[the remark this answers is not in the document]
Comment (typfel) [reply persistence/src/androidMain/kotlin/com/wire/kalium/persistence/db/Database.kt:10]: I'll let like this for now then, and update once #188 is merged.

**Pull request merged**